os_migrate/plugins/module_utils/server.py

```diff
@@ -128,9 +141,60 @@ def update_migration_params(self, params_dict):
             if value not in choices:
                 raise exc.UnexpectedChoice('floating_ip_mode', choices, value)
 
+        # TODO: add some validation for block device mapping v2
+        if 'data_copy' in params_dict:
+            value = params_dict['data_copy']
+            if not isinstance(value, bool):
+                raise exc.InvalidInputType('data_copy', 'boolean')
+
         return super().update_migration_params(params_dict)
 
-    def update_sdk_params_block_device_mapping(self, sdk_params, block_device_mapping):
+    def update_sdk_params_block_device_mapping_nocopy(self, sdk_params):
+        params, info = self.params_and_info()
+        migration_params = self.migration_params()
+        sdk_params['block_device_mapping'] = []
+        block_device_mapping = sdk_params['block_device_mapping']
+        image_id = sdk_params.get('image_id', None)
+        additional_volumes = migration_params['additional_volumes']
+
+        # Update block device mapping with volumes if data_copy is false
+        if migration_params['boot_volume']['uuid']:
+            boot_vol_mapping = {
+                "boot_index": 0,
+                "uuid": migration_params['boot_volume']['uuid'],
+                'delete_on_termination': False,
+                'destination_type': 'volume',
+                'source_type': 'volume',
+            }
+            block_device_mapping.append(boot_vol_mapping)
+
+        if additional_volumes is not None:
+            for volume in additional_volumes:
+                if volume['uuid']:
+                    additional_volume_mapping = {
+                        "boot_index": -1,
+                        'uuid': volume['uuid'],
+                        'delete_on_termination': False,
+                        'destination_type': 'volume',
+                        'source_type': 'volume',
+                    }
+                    block_device_mapping.append(additional_volume_mapping)
+            del sdk_params['image_id']
+        elif image_id is not None:
```

Comment:
The elif/else block belongs to `if migration_params['boot_volume']['uuid']`. All of that deals with how the machine boots. The `additional_volumes` block deals with optional volumes and is independent.